protocol/0087-VAMM-automated_market_maker.md

```diff
@@ -0,0 +1,94 @@
+
+- Placing orders directly on the book
+  - This approach entails the AMM generating orders at some frequency and placing them directly on the book. In this case for the rest of the system behaviour would be identical to if an external party were placing these orders and managing any resulting trades
+- A separate off-book liquidity source, which is queried after trades at each price level
+  - This approach entails the AMM acting as a separate liquidity pool which only acts when an aggressive order arrives. At that point in time, for each price level the protocol would first check the order book for volume, trading with that if it is available, it would then check for any AMMs offering volume at that price level, trading with those if available, before moving on to the next best price level in a similar manner. 
+
+
+For each, there are associated benefits and risks:
+
+- Placing orders directly on the book
+  - Benefits
+    - Outside the loop of generating and placing these orders the system needs very few changes, everything else can interact with and visualise the market as before
+  - Risks
+    - A large number of orders may need to be generated each time the AMM is updated. It is possible that this compounding with large numbers of people utilising an AMM could result in performance impacts.
+    - If the AMMs update too infrequently the spread on the market could widen (as opposed to a true AMM which updates immediately after each trade). A once-per-block update would mean that a buy immediately following someone else's sell would not benefit from the reduced price on the AMM
+- Separate off-book liquidity
+  - Benefits
+    - Only requires updating or interacting with on the arrival of orders which may immediately trade, improving performance
+    - Acts as a more 'pure' AMM structure, immediately updating on each trade before interacting with the next
+  - Risks
+    - The AMMs would have to be each checked for every tick on the market, adding a performance impact to the choice of decimal places, or another parameter which needed to be set for tick size of AMMs on a market. It is possible this performance impact is significant.
+      - This also requires that core matching code outside the new AMM logic is aware of these changes and can handle them, potentially adding uncertainty to scope of change work.
+    - A "complete" order book picture would now require understanding the AMM presence at a given timepoint too. To build a full order book, calculations would have to be performed in downstream systems to expand the AMM into virtual orders, combining those with real orders resting on the book. Failure to do this correctly would result in an inaccurate picture of available liquidity.
+    - Additional margin considerations for the AMM. When expanding out the AMM to virtual orders for a book, how does one ensure liquidity is not shown which would actually not be tradable with the funds available to that AMM.
```

Comment:
I think something like once per block or once per mark price update we'd have to calculate the virtual order volume at each price level and calculate the order margin based on that. Or we look if there is some easy approximation.